zcash_client_sqlite/src/lib.rs

```diff
@@ -1075,23 +1085,32 @@ impl<P: consensus::Parameters> WalletWrite for WalletDb<rusqlite::Connection, P>
                             Some(output.memo()),
                         )?;
 
-                        if matches!(recipient, Recipient::InternalAccount(_, _)) {
+                        if matches!(recipient, Recipient::InternalAccount { .. }) {
                             wallet::sapling::put_received_note(wdb.conn.0, output, tx_ref, None)?;
                         }
                     }
                     TransferType::Incoming => {
-                        match spending_account_id {
-                            Some(id) => {
-                                if id != *output.account() {
-                                    panic!("Unable to determine a unique account identifier for z->t spend.");
-                                }
-                            }
-                            None => {
-                                spending_account_id = Some(*output.account());
-                            }
-                        }
-
                         wallet::sapling::put_received_note(wdb.conn.0, output, tx_ref, None)?;
+
+                        if let Some(account_id) = funding_account {
+                            let recipient = Recipient::InternalAccount {
+                                receiving_account: *output.account(),
+                                // TODO: recover the actual UA, if possible
+                                external_address: Some(Address::Sapling(output.note().recipient())),
```

Comment:
How can we ensure we distinguish between account-internal and cross-account transfers, if we always set `external_address`?

Reply:
Same answer as above, this is the contract of `TransferType`.

Reply:
Here I now agree that this is doing the correct thing.